hw9: advanced oop/hw9.py

```diff
@@ -0,0 +1,212 @@
+        chat.recieve(self)
+
+
+class User:
+    def __init__(self, name, ability):
+        self.ability = ability
+        self.name = name
+
+    def greeting(self):
+        print(f'Hi, {self.name}! U r gorgeous')
+
+    def description(self):
+        print(f'{self.name} {self.ability}')
+
+
+# task 2: new syntax for Python
+
+# to avoid annoying nested list let's flatten them
+def flatten_list(nested_list):
+    flattened_list = []
+    for element in nested_list:
+        if isinstance(element, list):
+            flattened_list += flatten_list(element)
+        else:
+            flattened_list.append(element)
+    return flattened_list
+
+
+class Args:
+    def __init__(self, *args, **kwargs):
+        self.args = list(args)
+        self.args = flatten_list(self.args)
+        self.kwargs = list(kwargs.values())
+
+        self.all_argies = self.args + self.kwargs
+
+    def __rlshift__(self, other):
+        return other(*self.all_argies)
+
+
+# task 3: a new way to pass arguments to functions
+
+class StrangeFloat(float):
+    def __getattr__(self, attr):
+        if '__' + attr.split('_')[0][:3] + '__' in dir(float):
+            fnc = getattr(float, '__' + attr.split('_')[0][:3] + '__')
+            return fnc(self, float(attr.split('_')[1]))
+        else:
+            dict_attr = {'subtract': self - float(attr.split('_')[1]),
+                         'multiply': self * float(attr.split('_')[1]),
+                         'divide': self / float(attr.split('_')[1])}
+            return StrangeFloat(dict_attr[attr.split('_')[0]])
+
+        def __repr__(self):
+            return float(self)
+
+
+# task 4: rewrite changing to dunders
+
+# overall, 3 (three) dunders
+import numpy as np
+
+matrix2 = []
+for idx in range(0, 100, 10):
+    matrix2.__iadd__([list(range(idx, idx + 10))])
+
+selected_columns_indices = list(filter(lambda x: x in range(1, 5, 2), range(matrix.__len__())))
+selected_columns = map(lambda x: [x.__getitem__(col) for col in selected_columns_indices], matrix)
+
+arr = np.array(list(selected_columns))
+
+mask = arr[:, 1] % 3 == 0
+new_arr = arr[mask]
+
+product = new_arr @ new_arr.T
+
+if (product[0] < 1000).all() and (product[2] > 1000).any():
+    print(product.mean())
+
+
+# task 5:
+
+from abc import ABC, abstractmethod
+
+
+class BiologicalSequence(ABC):
+    def __init__(self, sequence):
+        self.sequence = sequence.upper()
+
+    def __len__(self):
+        return len(self.sequence)
+
+    @abstractmethod
+    def __getitem__(self, slc):
+        pass
+
+    @abstractmethod
+    def check_alphabet(self):
+        pass
+
+    def __repr__(self):
+        return str(self.sequence)
+
+
+class NucleicAcidSequence(BiologicalSequence):
+    def check_alphabet(self):
+        check_na = (['U'], ['T'])[int(self.sequence_type == "DNA")]
+        check_set = set(['A', 'G', 'C'] + check_na)
+        len_foreign = len(set(self.sequence) - check_set)
+        assert len_foreign == 0, 'wrong alphabet!!!'
```

Comment:
Здесь можно было сделать по аналогии с `complement`. И для РНК, и для ДНК задать специальные атрибуты с допустимым для них алфавитом и использовать его здесь
```suggestion
        len_foreign = len(set(self.sequence) - self.alphabet)
        assert len_foreign == 0, 'wrong alphabet!!!'
```